Check kernel params for bridge traffic filtering

miminar opened the pull request:
Check whether bridge traffic is passed to iptables' chains if started
with `--icc=false`. Inter-container communication can be restricted only
if following parameters are set.
- `/proc/sys/net/bridge/bridge-nf-call-iptables`
- `/proc/sys/net/bridge/bridge-nf-call-ip6tables`

Resolves issue #11404

LK4D4 commented on daemon/networkdriver/bridge/driver.go:
```diff
@@ -82,6 +91,74 @@ var (
+		enabled = true
+		err     error
+	)
+	if int(ipVer)&int(ipv4) > 0 {
```
I think, using `&` is pretty confusing. Why not use just `==`

miminar replied:
I'll get rid of bit operations.

LK4D4 commented on daemon/networkdriver/bridge/driver.go:
```diff
@@ -82,6 +91,74 @@ var (
+// on particular interface. Interface may be specified for IPv6 only. If
+// `iface` is empty, `default` will be assumed, which represents default value
+// for new interfaces.
+func getForwardingKernelParam(ipVer ipVersion, iface string) (res string) {
```
It's not recommended using named returns for less then 3 returning values.

LK4D4 commented on daemon/networkdriver/bridge/driver.go:
```diff
@@ -82,6 +91,74 @@ var (
+
+func isBridgeNetFilteringEnabled(ipVer ipVersion) (bool, error) {
+	var (
+		enabled = true
```
why enabled by default?

miminar replied:
For the same reason the `all()` returns `True` for `[]` in python:

    for cond in condList:
        if not cond:
             return False
    return True

I'll rewrite it though.

LK4D4 commented on daemon/networkdriver/bridge/driver.go:
```diff
@@ -82,6 +91,74 @@ var (
+func getBridgeNFKernelParam(ipVer ipVersion) (res string) {
+	switch ipVer {
+	case ipv4:
+		res = "/proc/sys/net/bridge/bridge-nf-call-iptables"
```
should be `return` everywhere in such cases.

miminar replied:
OK, got it.

LK4D4 commented on daemon/networkdriver/bridge/driver.go:
```diff
@@ -82,6 +91,74 @@ var (
+}
+
+func isPacketForwardingEnabled(ipVer ipVersion, iface string) (bool, error) {
+	var (
```
You don't need vars mostly. You can just return immediately when you get result.

thaJeztah commented:
@LK4D4 I think this PR: https://github.com/docker/docker/pull/11391 is trying to solve the same issue. I think **this** PR is more complete/mature. Perhaps the other one should be closed?

jamtur01 commented on docs/sources/articles/networking.md:
```diff
@@ -324,6 +324,15 @@ page for further details.
 > `--link=` option.
 
+> **Note**:
+> For `--icc=false` option to work, the bridge's traffic must be passed to
```
For the ...

jamtur01 commented on docs/sources/articles/networking.md:
```diff
@@ -324,6 +324,15 @@ page for further details.
 
+> **Note**:
+> For `--icc=false` option to work, the bridge's traffic must be passed to
+> iptables' chains. This is controled by kernel parameters
```
by the kernel

jamtur01 replied:
controlled

jamtur01 commented on docs/sources/articles/networking.md:
```diff
@@ -324,6 +324,15 @@ page for further details.
+> For `--icc=false` option to work, the bridge's traffic must be passed to
+> iptables' chains. This is controled by kernel parameters
+> `net.bridge.bridge-nf-call-iptables` and
+> `net.bridge.bridge-nf-call-ip6tables`. If set to 1, traffic will be processed
```
If they are set to 1...

jamtur01 commented on docs/sources/articles/networking.md:
```diff
@@ -324,6 +324,15 @@ page for further details.
+> `net.bridge.bridge-nf-call-ip6tables`. If set to 1, traffic will be processed
+> with `iptables` rules. However, several distributions set them to 0 by
+> default. Docker checks these parameters upon its initialization and will
+> refuse to start unless enabled if `--icc=false` is requested.
```
unless they are enabled? if the \`--icc-false\` option is requested. 

Although the whole sentence could do with a rewrite.

miminar commented:
@jamtur01, @LK4D4: Is it any better now?

jamtur01 commented on docs/sources/articles/networking.md:
```diff
@@ -324,6 +324,15 @@ page for further details.
+> `net.bridge.bridge-nf-call-iptables` and
+> `net.bridge.bridge-nf-call-ip6tables`. If they are set to 1, traffic will be
+> processed with `iptables` rules. However, several distributions set them to 0
+> by default. If the `--icc=false` is given, Docker will check these parameters
```
If the \`--icc=false\` flag is set,

icecrime commented:
Ping @aboch @mavenugo @mrjana: looking good to you?

Commit:
Check and set kernel params for bridge traffic filtering

Check whether bridge traffic is passed to iptables' chains if started
with `--icc=false`. Inter-container communication can be restricted only
if following parameters are set.

- `/proc/sys/net/bridge/bridge-nf-call-iptables`
- `/proc/sys/net/bridge/bridge-nf-call-ip6tables`

If not present, try to load `br_netfilter` kernel module or ask the user
to do it. If unset, try to set them.

Resolves issue #11404

Signed-off-by: Michal Minar <miminar@redhat.com>

miminar commented:
Thanks @aboch! Are your observations covered now?

mrjana commented:
@miminar A high level design question. In your PR you are proposing to fail daemon start if `--icc=false` and one of the bridge forward to iptables procentry is 0. Rather should we check and set these entries to 1 so that it works in `--icc=false` case rather than bailing out and having the user to make more changes? Wouldn't that be a much better user experience?

miminar commented:
@mrjana I don't like to mess with system configuration from inside of service. I'm aware that this how it's being done with forwarding kernel parameters. But perhaps admin has its reasons for setting it up this way. Docker init's side-effects may prove as quite surprising to such an admin.

But I'm not completely opposed to the idea. However, I'd at least accompany these system tweaks with some warning messages.

dave-tucker commented:
@mrjana @miminar these values can only be set if the `br_netfilter` module is loaded.
AFAIK, it was mainlined in 3.18-rc1 but may still require the user to `modprobe br_netfilter` or install additional packages (depending on their distro) and so I would discourage setting these automatically.

Overall, I'm :+1: to this PR, but I do recall that there were some issues with these values being set that were exposed [in OpenStack](http://buriedlede.blogspot.com.au/2012/07/debugging-networking-problems-with.html) but I should hope these have been resolved by now.

mrjana commented:
@dave-tucker these parameters are serviced by bridge module which is anyways a requirement for docker networking. But regardless, all I am saying is set it to 1 only if these entries are present and set to 0

@miminar docker already sets system level params like ip_forwarding automatically based on the intent of  a daemon configuration. So my thinking is it would be much better user experience to chain these automatically to iptables and achieve `--icc=false` rather then failing with error and a operator intervention being required. Docker is an infrastructure tool and so it is expected to modify certain system parameters. But if you are not comfortable with it that is fine too.

miminar commented:
@mrjana OK. I'll rework it so that parameters get set to keep behavior consistent with forwarding kernel params. Changing the system modification policy should go to another issue/PR.

dave-tucker commented:
@mrjana output from latest b2d...

``` bash
$ uname -a
Linux dev 3.18.5-tinycore64 #1 SMP Sun Feb 1 06:02:30 UTC 2015 x86_64 GNU/Linux

$ lsmod | grep bridge
bridge                 57344  0
stp                    12288  1 bridge
llc                    12288  2 bridge,stp
ipv6                  229376 29 bridge,[permanent]

$ ls /proc/sys/net/
core/             ipv6/             nf_conntrack_max
ipv4/             netfilter/        unix/

# Note: /proc/sys/net/bridge/ does not exist

$ sudo modprobe br_netfilter

$ ls /proc/sys/net/bridge/
bridge-nf-call-arptables       bridge-nf-call-iptables        bridge-nf-filter-vlan-tagged
bridge-nf-call-ip6tables       bridge-nf-filter-pppoe-tagged  bridge-nf-pass-vlan-input-dev

...it would appear that as I stated originally, these are tied to the `br_netfilter` module, *not* the `bridge` module. 

My concern here is that setting these parameters will fail on systems where the above stated module is not loaded:

``` bash
$ sudo sysctl -w bridge.bridge-nf-call-iptables=1
sysctl: error: 'bridge.bridge-nf-call-iptables' is an unknown key
```
We should ensure that an appropriate error message is given to the user in this instance (e.g "please ensure you have the br_netfilter kernel module loaded". In addition, I think an accompanying patch in to b2d add support for`br_netfilter` is required as well as mentioning this dependency in the docs.

/cc @minimar

miminar commented:
@dave-tucker sounds reasonable. I'll update PR tomorrow.

mrjana commented:
@dave-tucker What I was saying is that the code this PR brings in only takes effect if that sysctl parameter is present which can be present only (as you demonstrated) only when br_netfilter is loaded. All I am saying is that we should set it automatically only when it is already present and it is set to 0 and we need to enable `--icc=false`. This would mean we will never get into a situation where we are trying to set it if the module is not present.

miminar commented:
@dave-tucker @mrjana updated. Params are set during init. If they are missing, docker will spit a warning saying to load `br_netfilter`. I also mentioned the module in requirements. Maybe those warnings could be turned into errors, what do you thing?

calavera commented on daemon/networkdriver/bridge/driver_test.go:
```diff
@@ -31,6 +31,15 @@ func findFreePort(t *testing.T) string {
 
+func TestIPConstantValues(t *testing.T) {
+	if ipv4|ipv6 != ipvboth {
+		t.Fatal("bitwise or of ipv4(%#04b) and ipv6(%#04b) must yield ipvboth(%#04b)", ipv4, ipv6, ipvboth)
```
I think these two `t.Fatal` need to be `t.Fatalf`

miminar replied:
Good catch!

dave-tucker commented:
thanks for the changes @miminar, LGTM

miminar commented:
Ping @mrjana, @LK4D4, @icecrime, @aboch. Can we please move on with this? Is there anything I can improve?

cpuguy83 commented on daemon/networkdriver/bridge/driver.go:
```diff
@@ -105,6 +116,108 @@ type Config struct {
+	}
+}
+
+func enableBridgeNetFiltering(ipVer ipVersion, iface string) error {
```
If we're going to enable, should we modprobe `br_netfilter` first?

LK4D4 replied:
I thought it is writing error about modprobing.

cpuguy83 replied:
Looks like it's just checking if it's enabled, not making sure module is loaded.

LK4D4 replied:
It is checking error from this function.

LK4D4 replied:
It is checking error from this function.

miminar replied:
Modprobing it first seems reasonable. I'll add it.

jfrazelle commented:
LGTM

jfrazelle commented:
ping @LK4D4

LK4D4 commented:
I'll test this soon. Code looks good.

LK4D4 commented:
@miminar I have no `/proc/sys/net/bridge/bridge-nf-call-iptables` file and no `br_netfilter` module loaded and I can run my daemon with `--icc=false` without any errors or warnings and my containers can talk to each other.

LK4D4 commented on daemon/networkdriver/bridge/driver.go:
```diff
@@ -225,17 +338,33 @@ func InitDriver(config *Config) error {
+			logrus.Warnf("Unable to enable bridge traffic filtering: %v", err)
+			if ptherr, ok := err.(*os.PathError); ok {
+				if errno, ok := ptherr.Err.(syscall.Errno); ok && errno == syscall.ENOENT {
+					logrus.Warnf("Cannot restrict inter-container communication. Please ensure that br_netfilter kernel module is loaded.")
```
I think we should return error here.

miminar replied:
I agree. That's how it was originally - only without Docker trying to correct the settings. Now when Docker attempts to do anything possible to make it right, I think it's safe to make it an error.

GordonTheTurtle commented:
Can you please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "check-kernel-params" git@github.com:miminar/docker.git somewhere
$ cd somewhere
$ git commit --amend -s --no-edit
$ git push -f
~~~

This will update the existing PR, so **DO NOT** open a new one.

LK4D4 commented:
@miminar Tahnks, seems like we have bug which your code revealed.

LK4D4 commented on daemon/networkdriver/bridge/driver.go:
```diff
@@ -242,18 +355,48 @@ func InitDriver(config *Config) error {
-			if err := ioutil.WriteFile("/proc/sys/net/ipv6/conf/all/forwarding", []byte{'1', '\n'}, 0644); err != nil {
-				logrus.Warnf("WARNING: unable to enable IPv6 all forwarding: %s\n", err)
+		}
+		if tryLoad {
```
I don't understand why we need this block, it is too late here to load module. You should just throw error or load module before enabling filtering.

miminar replied:
`br_netfilter` may be compiled-in (kernels before 3.18.0). Modprobing it before checking for existence of `/proc/sys/net/bridge/*` would generate needless warnings on these kernels.

I'm not opposed to modprobing module before trying to enable filtering, but it would still require a check for existence of those params (which is now handled by the first call to `enableBridgeNetFiltering()`). Is it okay to leave modprobing code in [if !config.InteContainerCommunication ...](https://github.com/docker/docker/pull/11405/files#diff-fb93346b0b06f64fcb85ea4a7de12f8dR373) block?

LK4D4 commented:
ping @mavenugo 
Somehow there is no `/proc/sys/net/bridge` inside testing container even if br_netfilter loaded.

mavenugo commented:
@LK4D4 we are not mounting `/proc/sys` no the testing container. Should we mount that in the `Makefile DOCKER_MOUNT` ?

LK4D4 commented:
@mavenugo But other networking options like forwarding there :/ Also we need some teardown for this then.

miminar commented:
@mavenugo @LK4D4 are you looking into mounting those params or shall I take a stab at it?

LK4D4 commented:
Okay, now it seems that we need that stuff always.

jfrazelle commented:
@miminar can you update the PR to add the modprobe for br_netfilter

jfrazelle commented:
otherwise one of us can carry, but we need this :)

miminar commented:
@jfrazelle I'm afraid it's not that simple. `br_netfilter` is loaded inside a container if it's loaded on the host. Only the `/proc/sys/net/bridge` directory isn't available. And it cannot be mounted with `-v /proc/sys/net/bridge:/proc/sys/net/bridge` - entries cannot be created in `proc` filesystem with `touch` or `mkdir`.

This seems to me like a shorting in kernel's network namespacing. It properly hides configuration entries for host's interfaces from inside container and creates new ones for it. IMHO it should create new container entries for `/proc/sys/net/bridge/*` in similar way as for `/proc/sys/net/ipv4/*`.

Now we can either disable bridge settings checking if running from inside of container or somehow mock kernel param paths for testing purposes. What do you think?

jfrazelle commented:
Well we need to load it on the host is what I am saying, like how we do
with aufs and other modules

On Tuesday, May 12, 2015, Michal Minar <notifications@github.com> wrote:

> @jfrazelle <https://github.com/jfrazelle> I'm afraid it's not that
> simple. br_netfilter is loaded inside a container if it's loaded on the
> host. Only the /proc/sys/net/bridge directory isn't available. And it
> cannot be mounted with -v /proc/sys/net/bridge:/proc/sys/net/bridge -
> entries cannot be created in proc filesystem with touch or mkdir.
>
> This seems to me like a shorting in kernel's network namespacing. It
> properly hides configuration entries for host's interfaces from inside
> container and creates new ones for it. IMHO it should create new container
> entries for /proc/sys/net/bridge/* in similar way as for
> /proc/sys/net/ipv4/*.
>
> Now we can either disable bridge settings checking if running from inside
> of container or somehow mock kernel param paths for testing purposes. What
> do you think?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/11405#issuecomment-101192312>.
>

miminar commented:
That doesn't help. `/proc/sys/net/bridge` isn't available inside development container due to network namespacing regardless of `br_netfilter` (un)loaded on host.

jfrazelle commented:
Ya but what I'm saying is, sorry and I can just make a separate pr for it,
we need to load it on the host, we were seeing it not being loaded and
communication between containers is broken w hairpin nat without it

On Tuesday, May 12, 2015, Michal Minar <notifications@github.com> wrote:

> That doesn't help. /proc/sys/net/bridge isn't available inside
> development container due to network namespacing regardless of
> br_netfilter (un)loaded on host.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/11405#issuecomment-101309585>.
>

miminar commented:
@jfrazelle got it! Sorry it took me so long :-)

miminar commented:
Rebased. Also added a check for running inside of Docker container. In such a case Docker won't stop on missing kernel params.

cpuguy83 commented:
ping @mavenugo @mrjana 
This is touching some code that has been moved to libnetwork.

mrjana commented:
@miminar Is it possible to port this to `docker/libnetwork`?

icecrime commented:
@minimar Sorry I'm going to close this as it can't be merged because of libnetwork integration. I'll leave it to you and @mrjana to figure if you are available to port it, or if the networking guys are available to carry. Thanks, and sorry for this complication.
